### PR TITLE
Introduce slashing/jailing for DPoS

### PR DESCRIPTION
**Pull request by s2quake**
### Features

* Slashing -> SlashCtrl.Shash
* Jailing -> ValidatorCtrl.Jail
* Unjailing -> ValidatorCtrl.Unjail
* Tombstone -> ValidatorCtrl.Tombstone
* Penalties for double signing -> EvidenceCtrl

### Test Codes

* SlashCtrlTest: Unit and senario test
* ValidatorSigningInfoCtrlTest: Unit test

### References

* There is code(from cosmos) for infraction due to network latency, but it has not been tested.
* The code for the case where the slash factor is 100% is not implemented.
* The test code for the case where there is a remainder when dividing by a slash factor is not implemented.
* Very complex test code like delegate->redelegate->adelegate is not implemented.

### CODE REVIEW

File: Lib9c/Action/DPoS/Control/SlashCtrl.cs [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+
+            if (infractionHeight > actionContext.BlockIndex)
+            {
+                throw new ArgumentOutOfRangeException(
+                    "impossible attempt to slash future infraction",
+                    nameof(infractionHeight));
+            }
+            else if (infractionHeight == actionContext.BlockIndex)
+            {
+            }
```
Comment (riemannulus) [inline Lib9c/Action/DPoS/Control/SlashCtrl.cs:57]: is it necessary?

File: Lib9c/Action/DPoS/Control/SlashCtrl.cs [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+                var unbondingSet = UnbondingSetCtrl.GetUnbondingSet(world)!;
+                foreach (var item in unbondingSet.UndelegationAddressSet)
+                {
+                    var undelegation = UndelegateCtrl.GetUndelegation(world, item)!;
+                    if (undelegation.ValidatorAddress == validatorAddress)
+                    {
+                        world = SlashUnbondingDelegation(world, actionContext, undelegation, infractionHeight, slashFactor, out var amountSlashed);
+                        remainingSlashAmount -= amountSlashed;
+                    }
+                }
```
Comment (riemannulus) [inline Lib9c/Action/DPoS/Control/SlashCtrl.cs:69]: I recommend separating this part into a separate method to make it more semantic.

File: Lib9c/Action/DPoS/Control/SlashCtrl.cs [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+
+                foreach (var item in unbondingSet.RedelegationAddressSet)
+                {
+                    var redelegation = RedelegateCtrl.GetRedelegation(world, item)!;
+                    if (redelegation.SrcValidatorAddress == validatorAddress)
+                    {
+                        world = SlashRedelegation(world, actionContext, redelegation, infractionHeight, slashFactor, nativeTokens, out var amountSlashed);
+                        remainingSlashAmount -= amountSlashed;
+                    }
+                }
```
Comment (riemannulus) [inline Lib9c/Action/DPoS/Control/SlashCtrl.cs:79]: I recommend separating this part into a separate method to make it more semantic.

File: Lib9c/Action/DPoS/Control/SlashCtrl.cs (hunk 1 above)
Comment (s2quake) [reply Lib9c/Action/DPoS/Control/SlashCtrl.cs:57]: No, This code is a remnant of C...

File: Lib9c/Action/DPoS/Control/SlashCtrl.cs (hunk 2 above)
Comment (s2quake) [reply Lib9c/Action/DPoS/Control/SlashCtrl.cs]: Separated code into the 'SlashUndelegations' method.

File: Lib9c/Action/DPoS/Control/SlashCtrl.cs (hunk 3 above)
Comment (s2quake) [reply Lib9c/Action/DPoS/Control/SlashCtrl.cs]: Separated code into the 'SlashRedelegations' method.

File: Lib9c/Action/DPoS/Model/RedelegationEntry.cs [hunk 4]
```diff
@@ -26,10 +26,12 @@ public class RedelegationEntry : IEquatable<RedelegationEntry>
             Address = DeriveAddress(redelegationAddress, index);
             RedelegationAddress = redelegationAddress;
             RedelegatingShare = redelegatingShare;
+            InitialConsensusToken = unbondingConsensusToken;
             UnbondingConsensusToken = unbondingConsensusToken;
             IssuedShare = issuedShare;
             Index = index;
             CompletionBlockHeight = blockHeight + UnbondingSet.Period;
```
Comment (OnedgeLee) [inline Lib9c/Action/DPoS/Model/RedelegationEntry.cs:33]: Unbonding period have to be maintained with protective manner, to find effective block height easily for production, IMO.

File: Lib9c/Action/DPoS/Control/SlashCtrl.cs [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+                if (entry.CreationHeight < infractionHeight)
+                {
+                    continue;
+                }
+                if (entry.IsMatured(infractionHeight))
+                {
+                    continue;
+                }
+
+                var (q, r) = entry.InitialConsensusToken.DivRem(slashFactor);
```
[the remark this answers is not in the document]
Comment (OnedgeLee) [reply Lib9c/Action/DPoS/Control/SlashCtrl.cs:254]: It follows original spec, closing

File: Lib9c/Action/DPoS/Model/RedelegationEntry.cs (hunk 4 above)
Comment (s2quake) [reply Lib9c/Action/DPoS/Model/RedelegationEntry.cs:33]: I think we should work on `Unbonding period` in a different PR.

**Review by riemannulus — APPROVED** [review]

**Pull request merged**